src/deploy/functions/containerCleaner.ts

```diff
@@ -128,6 +133,196 @@ export class ContainerRegistryCleaner {
+    throw new FirebaseError("Invalid region supplied.");
+  }
+
+  const artifacts = new Set<string>();
+  const failedSubdomains = new Set<string>();
+  const subdomains = getSubdomainsFromLocations(locations);
+  for (const subdomain of subdomains) {
+    if (dockerHelpers[subdomain] === undefined) {
+      const origin = `https://${subdomain}.${containerRegistryDomain}`;
+      dockerHelpers[subdomain] = new DockerHelper(origin);
+    }
+
+    const searchPaths = findGCFPaths(projectId, subdomain, locations);
+    for (const path of searchPaths) {
+      try {
+        const rootChildren = (await dockerHelpers[subdomain].ls(path)).children;
+        if (rootChildren.length === 0) {
+          continue;
+        }
+        // partition children into subsets that are artifacts and location dirs
+        const locationDirs = rootChildren.filter((child) => SUBDOMAIN_MAPPING[child] !== undefined);
+        const artifactCandidates = rootChildren.filter(
+          (child) => SUBDOMAIN_MAPPING[child] === undefined
+        );
+        // search 1 level deeper on directories
+        for (const dir of locationDirs) {
+          try {
+            const innerPath = path + `/${dir}`;
+            const innerChildren = (await dockerHelpers[subdomain].ls(innerPath)).children;
+            await formatTagsAndInsert(
+              dockerHelpers[subdomain],
+              innerChildren,
+              innerPath,
+              artifacts
+            );
+          } catch (err) {
+            logger.debug(err);
+          }
+        }
+        // format the artifacts with tags
+        try {
+          await formatTagsAndInsert(dockerHelpers[subdomain], artifactCandidates, path, artifacts);
+        } catch (err) {
+          logger.debug(err);
+        }
+      } catch (err) {
+        failedSubdomains.add(subdomain);
+        logger.debug(err);
+      }
+    }
+  }
+
+  if (failedSubdomains.size == SUBDOMAINS.length) {
+    throw new FirebaseError("Failed to search all subdomains.");
+  } else if (failedSubdomains.size > 0) {
+    const failed: string = [...failedSubdomains].join(",");
+    throw new FirebaseError(`Failed to search the following subdomains: ${failed}`);
+  }
+
+  return artifacts;
+}
+
+/**
+ * Deletes all artifacts from GCF directory in GCR.
+ * @param projectId: the current project that contains GCF artifacts
+ * @param location: the specific region to be clean up. If omitted, will delete all locations.
+ * @param dockerHelpers: a map of {@link SUBDOMAINS} to {@link DockerHelper}. If omitted, will use the default value and create each {@link DockerHelper} internally.
+ *
+ * @throws {@link FirebaseError}
+ * Thrown if the provided location is not a valid Google Cloud region or we fail to delete subdomains.
+ */
+export async function deleteGCFArtifacts(
+  projectId: string,
+  locations?: string[],
+  dockerHelpers: Record<string, DockerHelper> = {}
```

Comment:
Same here - This is using a default value so we can inject a stub for unit testing.